Mage/src/main/java/mage/players/PlayerImpl.java

```diff
@@ -821,7 +821,7 @@ the moment before the cost was paid (see rule 717, "Handling Illegal Actions").
         // So discard is also successful if card is moved to another zone by replacement effect!
         game.fireEvent(GameEvent.getEvent(GameEvent.EventType.DISCARDED_CARD, card.getId(), source, playerId));
 
-        if (fireFinalEvent) {
+        if (payForCost) {
             game.fireEvent(GameEvent.getEvent(GameEvent.EventType.DISCARDED_CARDS, null, source, playerId, 1));
```

Comment:
That's wrong. If you want to check event for discard cost or discard effect then use event's flag in the bugged ability. Example:
![shot_210124_203358](https://user-images.githubusercontent.com/8344157/105636822-80d37300-5e83-11eb-9e11-1126962460ed.png)

Even more a `fireFinalEvent` was used for that special case:
![shot_210124_203131](https://user-images.githubusercontent.com/8344157/105636866-ac565d80-5e83-11eb-880e-4bd866c983b6.png)

I'm found that it was changed few commits ago by error. It must be used here as `true`:
![shot_210124_204653](https://user-images.githubusercontent.com/8344157/105637181-892cad80-5e85-11eb-99d1-a18dd5083f70.png)

So what you need:
* revert changes and return fire flag;
* fix default `discard(` to use true flag;
* find all `DISCARDED_CARDS` event usage and setup the event flag same way as normal event do (see screenshot above about cost/effect);
* change Rielle code to check that event's flag;
* add additional commands/checks in all of your tests:
![shot_210124_204142](https://user-images.githubusercontent.com/8344157/105637038-b0cf4600-5e84-11eb-9f87-e2c791f334b5.png)

Reply:
Seems that fixing default discard flag has solved the problem with Rielle. I've changed tests to include strict choose mode. Thanks for your help.